scheduler/src/cook/mesos/util.clj

```diff
@@ -298,65 +298,59 @@
+         (map (partial d/entity db))
+         (filter #(<= (.getTime start) (.getTime (:job/submit-time %))))
+         (filter #(< (.getTime (:job/submit-time %)) (.getTime end)))
+         (filter #(= state-keyword (:job/state %)))
+         (filter #(not (:job/custom-executor %))))))
+
+;; This differs from get-active-jobs-by-user-and-state as it is also looking up based on task state.
 (defn get-completed-jobs-by-user
   "Returns all completed job entities for a particular user
-   in the specified timeframe, without a custom executor."
+   in the specified timeframe, without a custom executor. Supports looking up based
+   on task state 'success' and 'failed' if passed into 'state'"
   [db user start end limit state name-filter-fn]
   (timers/time!
     get-completed-jobs-by-user-duration
-    (let [;; Expand the time range so that clock skew between cook
-          ;; and datomic doesn't cause us to miss jobs
-          ;; 1 hour was picked because a skew larger than that would be
-          ;; suspicious
-          expanded-start (Date. (- (.getTime start)
-                                   (-> 1 t/hours t/in-millis)))
-          expanded-end (Date. (+ (.getTime end)
-                                 (-> 1 t/hours t/in-millis)))
-          jobs
-          (->> (d/seek-datoms db :avet :job/user user (d/entid-at db :db.part/user expanded-start))
-               (take-while #(and (< (.e %) (d/entid-at db :db.part/user expanded-end))
-                                 (= (.a %) (d/entid db :job/user))
-                                 (= (.v %) user)))
-               (map #(.e %))
-               (map (partial d/entity db))
-               (filter #(<= (.getTime start) (.getTime (:job/submit-time %))))
-               (filter #(< (.getTime (:job/submit-time %)) (.getTime end)))
-               (filter #(= :job.state/completed (:job/state %)))
-               (filter #(not (:job/custom-executor %))))]
-      (->>
-        (cond->> jobs
-                 (instance-states state) (filter #(= state (job-ent->state %)))
-                 name-filter-fn (filter #(name-filter-fn (:job/name %))))
-        (take limit)))))
-
-;; This is a separate query from completed jobs because most running/waiting jobs
-;; will be at the end of the time range, so the query is somewhat inefficent.
-;; The standard datomic query performs reasonably well on the smaller set of
-;; running and waiting jobs, so it's implemented that way to keep things simple.
+    (->>
+      (cond->> (get-jobs-by-user-and-state-and-submit db user start end :job.state/completed)
+               (instance-states state) (filter #(= state (job-ent->state %)))
+               name-filter-fn (filter #(name-filter-fn (:job/name %))))
+      (take limit))))
+
 (defn get-active-jobs-by-user-and-state
   "Returns all jobs for a particular user in the specified state
    and timeframe, without a custom executor.
    Note that this query is not performant for completed jobs, use
```

Comment:
> Note that this query is not performant for completed jobs

Aren't we basically using the same query for completed jobs as we use here?

Reply:
Yes. Stale comment. Reviewed and other comments in this PR seem good.